src/track/external-captions-handler.js

```diff
@@ -0,0 +1,397 @@
+      });
+    } else {
+      let _activeCuesChanged = false;
+      for (let _activeTextCuesIndex = 0; _activeTextCuesIndex < this._player.activeTextCues.length; _activeTextCuesIndex++) {
+        let _cue = this._player.activeTextCues[_activeTextCuesIndex];
+        if (this._player.currentTime < _cue.startTime || _cue.endTime < this._player.currentTime) {
+          this._player.activeTextCues.splice(_activeTextCuesIndex, 1);
+          _activeCuesChanged = true;
+        }
+      }
+      while (this._player.currentTime > track.cues[this._externalCueIndex].startTime) {
+        this._player.activeTextCues.push(track.cues[this._externalCueIndex]);
+        this._externalCueIndex++;
+        _activeCuesChanged = true;
+      }
+      if (_activeCuesChanged) {
+        this._player.onCueChange({
+          payload: {
+            cues: this._player.activeTextCues
+          }
+        });
+      }
+    }
+  }
+
+  /**
+   * updating the index of the text cues to the right location after a user seeked.
+   * @returns {void}
+   * @private
+   */
+  _maybeSetExternalCueIndex(): void {
+    const textTrack = this._player.getTracks(TrackType.TEXT).filter(track => track.active && track.external)[0];
+    if (textTrack && textTrack.external) {
+      let cues = textTrack.cues;
+      let i = 0;
+      while (this._player.currentTime > cues[i].startTime && i < cues.length) {
+        i++;
+      }
+      this._externalCueIndex = i;
+    }
+  }
+
+  /**
+   * adding external tracks (native or moduled tracks)
+   * @returns {Promise<any>} returns a promise when all the tracks were parsed and added.
+   * @public
+   */
+  addTracks(): void {
+    const captions = this._player.config.sources.captions;
+    if (!captions) {
+      return;
+    }
+    let textTracks = this._player.getTracks(TrackType.TEXT);
+    let textTracksLength = textTracks.length || 0;
+    captions.forEach(caption => {
+      let track = new TextTrack({
+        active: false,
+        index: textTracksLength++,
+        kind: "subtitles",
+        label: caption.label,
+        language: caption.language,
+        external: true,
+        url: caption.url,
+        type: caption.type,
+        cues: []
+      });
+      let sameLangTrack = textTracks.find(textTrack => caption.language === textTrack.language);
+      // replace the track with a new one, update only the cues to []
+      if (sameLangTrack) {
+        this._player.setTextTrackCues(track);
+      } else {
+        this._player.addTrack(track);
+      }
+      if (this._player.config.playback.useNativeTextTrack)
+        this._setNativeTextTrack(track);
+    });
+  }
+
+  /**
+   * selects external track start listenning to cues
+   * @param {TextTrack} textTrack - selected text track
+   * @returns {void}
+   * @public
+   */
+  selectTextTrack(textTrack: TextTrack): Promise<*> {
+    return new Promise((resolve) => {
+      if (textTrack.cues.length > 0 && !this._player.config.playback.useNativeTextTrack) {
```

Comment:
if you follow the comment about managing `_cuesInProcessMap` (which should be `_textTrackStatus`) then you can just check for the status if it's `READY` and set it instead of checking its cues